dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/program/jdbc/JdbcProgramOrgUnitAssociationsStore.java

```diff
@@ -64,22 +75,70 @@
 
     private final JdbcTemplate jdbcTemplate;
 
-    public SetValuedMap<String, String> getOrganisationUnitsAssociations( Set<String> uids )
+    private final CacheProvider cacheProvider;
+
+    private final Environment env;
+
+    private Cache<Set<String>> programOrgUnitAssociationCache;
+
+    @PostConstruct
+    public void init()
     {
+        programOrgUnitAssociationCache = cacheProvider.newCacheBuilderForSet( String.class )
+            .forRegion( "pgmOrgUnitAssocCache" )
+            .expireAfterWrite( 1, TimeUnit.HOURS )
+            .withInitialCapacity( 100 )
+            .withMaximumSize( SystemUtils.isTestRun( env.getActiveProfiles() ) ? 0 : 1000 )
+            .build();
+    }
 
-        return jdbcTemplate.query(
-            buildSqlQueryForRawAssociation( uids ),
-            resultSet -> {
-                SetValuedMap<String, String> setValuedMap = new HashSetValuedHashMap<String, String>();
-                while ( resultSet.next() )
-                {
-                    setValuedMap.putAll(
-                        resultSet.getString( 1 ),
-                        Arrays.asList( (String[]) resultSet.getArray( 2 ).getArray() ) );
-
-                }
-                return setValuedMap;
-            } );
+    @EventListener
+    public void handleApplicationCachesCleared( ApplicationCacheClearedEvent event )
+    {
+        programOrgUnitAssociationCache.invalidateAll();
+    }
+
+    public SetValuedMap<String, String> getOrganisationUnitsAssociations( Set<String> uids )
+    {
+        SetValuedMap<String, String> setValuedMap = new HashSetValuedHashMap<String, String>();
```

Comment:
I think this code could have been made slightly easier to read and perform slightly better by:

1. Split the given program identifiers into two sets, one set for the program IDs which have cached values/org units and one set for the IDs which don't.
2. Filter the second database query to include only the ones which do not have cached values. Populate the cache accordingly. 

That way we don't have to re-fetch already cached org unit associations.

In any case, the code looks good, and I understand this code has been tested already, so let's go ahead and merge it. Good work.

Reply:
@ameenhere Can we track the above suggestions by Lars' for inclusion in the next patch release (i.e. dev branch), please.

Reply:
@larshelge  I had the same thought process initially. But in the end I resolved to this implementation because 
1. There is a db call anyway, and the overhead required to fetch "all requested program orgUnit associations" instead of "only uncached program orgUnit associations" would not be significant. And it gives a chance to update the cache entries. So went ahead with this. 
2. At the moment, this call is only used with "one program" always. Since its only used in addEnrollment and that always needs to check only orgUnit association for a single program.

Reply:
Alright, thanks @ameenhere , sounds good. I guess we could just change the method signature to accept one program and simplify the code, if that is what we need to in the foreseeable future?